### PR TITLE
Initial pass at storing symbolic factorization

### PR DESCRIPTION
**Pull request by Wimmerer**
This implementation dives a little deep into the guts of UMFPACK, but I suppose that's probably inevitable. I don't believe there's a good way to check that pattern(A1) == pattern(A2), so it just throws right now.

### CODE REVIEW

File: src/factorization.jl [hunk 1]
```diff
@@ -38,7 +38,7 @@ end
 struct UMFPACKFactorization <: AbstractFactorization
```
Comment (ChrisRackauckas) [inline src/factorization.jl:38]: can we add a bool for whether to redo the symbolic factorization?

Comment (ChrisRackauckas) [reply src/factorization.jl:38]: Default to no redoing it.

Comment (Wimmerer) [reply src/factorization.jl:38]: Where do you want that? In the algorithm? So ` UMFPACKFactorization(reuse_symbolic=false)`?

File: src/factorization.jl [hunk 2]
```diff
@@ -35,10 +35,11 @@ function init_cacheval(alg::LUFactorization, A, b, u)
 end
 
 # This could be a GenericFactorization perhaps?
-struct UMFPACKFactorization <: AbstractFactorization
+Base.@kwdef struct UMFPACKFactorization <: AbstractFactorization
+    reuse_symbolic::Bool = false
```
Comment (ChrisRackauckas) [inline src/factorization.jl:39]:
```suggestion
    reuse_symbolic::Bool = true
```

**Review by ChrisRackauckas — APPROVED** [review]

**Comment by ChrisRackauckas** [comment]
CI has a shit ton on it right now, so I ran it locally and it passed.

**Pull request merged**